prow/cmd/deck/main.go

```diff
@@ -1607,3 +1617,65 @@ func defaultLensRemoteConfig(lfc *config.LensFileConfig) error {
 
 	return nil
 }
+
+// validateStoragePath "validates" the storage bucket, *if* the `Deck.RestrictStoragePaths` config field is true.
+// The path should match a generic "<key-type>/<bucket>/<folder>/<sub-folder...>" pattern.
+// Additionally, the path's bucket should match either the built-in and default storage locations _or_
+// should be contained in the `Deck.AdditionalAllowedBuckets` config field.
+// If the `Deck.RestrictStoragePaths` field is false, no validation errors are returned (but there will be warnings logged).
+func validateStoragePath(cfg config.Getter, path string) error {
+	validate := func(cfg config.Getter, path string) error {
+		parts := strings.SplitN(path, "/", 4)
+		if len(parts) < 4 {
+			return httpError{
+				error:      fmt.Errorf("invalid path: %s expected <key-type>/<bucket>/<folder>/<sub-folder...>", path),
+				statusCode: http.StatusBadRequest,
+			}
+		}
+
+		// Check bucket
+		bucket := parts[1]
+		allowed_buckets := sets.NewString(cfg().Deck.AdditionalAllowedBuckets...)
+		for _, dc := range cfg().Plank.DefaultDecorationConfigs {
+			allowed_buckets.Insert(dc.GCSConfiguration.Bucket)
+		}
+		if !allowed_buckets.Has(bucket) {
+			return httpError{
+				error:      fmt.Errorf("bucket %q not in allowed list %v", bucket, allowed_buckets),
+				statusCode: http.StatusBadRequest,
+			}
+		}
+
+		return nil
+	}
+	actualErr := validate(cfg, path)
```

Comment:
Nit: Why is this named `validate`? Also a nit but I personally find code easier to read without inline funcs because with inline funcs you have to read all the logic rather than being able to say `okay we validate, if there was an error we do X etc`

Reply:
Sorry, refactoring noise. I'll take another pass. (Not quite satisfied with where I left this, either.)